File: WebWindowNetCore/Bounds.fs
namespace WebWindowNetCore
open System
open System.IO
open FSharpTools
open FSharpTools.Functional

type Bounds = {
    X: int option
    Y: int option
    Width: int option
    Height: int option
    IsMaximized: bool
}

module internal Bounds = 
    let private getPath id = 
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)
        |> Directory.attachSubPath id
        |> sideEffectIf (fun p-> Directory.Exists(p) = false) Directory.CreateDirectory
        |> Directory.attachSubPath "bounds.json"

    let retrieve id =
        getPath id
        |> File.readAllText 
        |> Option.defaultValue ""
        |> sideEffect (printfn  "%s")
        |> TextJson.deserialize<Bounds>
    
    let save id bounds =
        getPath id
        |> File.writeAllText (TextJson.serialize<Bounds> bounds) 

File: WebWindowNetCore/Bounds.cs
using CsTools;
using CsTools.Extensions;

namespace WebWindowNetCore;

record Bounds(int? X, int? Y, int? Width, int? Height, bool IsMaximized)
{
    public static Bounds Retrieve(string id)
        => GetPath(id)
            .ReadAllTextFromFilePath()
            .SideEffect(Console.WriteLine)
            ?.Deserialize<Bounds>(Json.Defaults) 
            ?? new(null, null, null, null, false);

    public static void Save(string id, Bounds bounds)
        => GetPath(id).WriteAllTextToFilePath(bounds.Serialize(Json.Defaults));

    static string GetPath(string id)
        => Environment
            .GetFolderPath(Environment.SpecialFolder.ApplicationData)
            .AppendPath(id)
            .SideEffect(d => d.EnsureDirectoryExists())
            .AppendPath("bounds.json");
}